vshard/replicaset.lua

```diff
@@ -75,6 +75,108 @@ local fiber_cond_wait = util.fiber_cond_wait
+        conn.vconnect.is_validated = true
+    end
+end
+
+--
+-- Initialize future object. Should be done, when connection is established
+-- (inside netbox_on_connect).
+--
+local function conn_init_vconnect(conn)
+    local vconn = conn.vconnect
+    if vconn and vconn.is_named then
+        -- Validation should be done after every reconnect.
+        vconn.is_validated = false
+        local opts = {is_async = true}
+        vconn.future = conn:call('vshard.storage._call', {'info'}, opts)
+        vconn.future_cond:broadcast()
+    end
+end
+
+--
+-- Check, that future is ready, and its result is expected, close the connection
+-- if the result is not correct. The function doesn't yield.
+-- @retval true; The correct response is received.
+-- @retval nil, ...; Response is not received or validation error happened.
+--
+local function conn_validate_vconnect(conn)
+    local vconn = conn.vconnect
+    -- conn.vconnect may be nil, if connection was created on old version
+    -- and the storage was reloaded to a new one.
+    if not vconn or vconn.is_validated then
+        return true
+    end
+    -- Nothing to do, but wait in such case.
+    if not vconn.future or not vconn.future:is_ready() then
+        return nil, lerror.vshard(lerror.code.CONNECTION_IS_NOT_ESTABLISHED,
+                                  conn.replica.id)
+    end
+    -- Critical errors. Connection should be closed after these ones.
+    local result, err = vconn.future:result()
+    if not result then
+        conn:close()
+        -- Failed to get response. E.g. access error.
+        return nil, lerror.make(err)
+    end
+    if result[1].name ~= conn.replica.name then
+        conn:close()
+        return nil, lerror.vshard(lerror.code.INSTANCE_NAME_MISMATCH,
+                                  conn.replica.name, result[1].name)
+    end
+    -- Don't validate until reconnect happens.
+    vconn.is_validated = true
+    return true
+end
+
+--
+-- Wait until the future object is ready. Returns remaining timeout.
+-- @retval timeout; Future boject is ready.
+-- @retval nil, err; Timeout passed.
+--
+local function conn_wait_vconnect(conn, timeout)
+    local vconn = conn.vconnect
+    -- Fast path. In most cases no validation should be done at all.
+    -- conn.vconnect may be nil, if connection was created on old version
+    -- and the storage was reloaded to a new one.
+    if not vconn or vconn.is_validated or
+       (vconn.future and vconn.future:is_ready()) then
+        return timeout
+    end
+    local deadline = fiber_clock() + timeout
+    -- Wait for connection to be established.
+    if not vconn.future and
+       not fiber_cond_wait(vconn.future_cond, timeout) then
+        return nil, lerror.timeout()
+    end
+    timeout = deadline - fiber_clock()
+    -- Wait for async call to return.
+    local res, err = util.future_wait(vconn.future, timeout)
+    if res == nil then
+        return nil, lerror.make(err)
```

Comment:
Same as before - can set `storage._call` to nil and make a synchronous request to check that. The previous occurrence could be tested with an `is_async = true` request.

Reply:
Added